### DISCUSSION TITLE
Read and display voltage from GPIO

### DISCUSSION THREAD

**Discussion by elik745i** [HASwitchPlate/openHASP#564] [Q&A]
## Questions
Is there possibility or any example how to read analog data from GPIO on esp32 with installed openHASP?
I would like to display battery level of the battery operated device and implement shutdown on level below certain value.

**Comment by fvanroie** [comment]
Check out the Custom Code section. You need to add that code yourself:
https://www.openhasp.com/0.7.0/firmware/compiling/custom/

**Comment by elik745i** [comment]
got it, in case if someone needs:, voltage divider hooked to GPIO35 and illumination sensor on GPIO34. Voltage below 15% going to heartbeat led on GPIO4

my_custom.cpp file:



/* MIT License - Copyright (c) 2019-2022 Francis Van Roie
   For full license information read the LICENSE file in the project folder */

// USAGE: - Copy this file and rename it to my_custom.cpp
//        - Change false to true on line 9

#include "hasplib.h"

#if defined(HASP_USE_CUSTOM) && true // <-- set this to true in your code

#include "hasp_debug.h"
#include "custom/my_custom.h"

unsigned long last_blink = 0;
bool blink_state = LOW;

const int LED_BUILTIN1 = 4;  //red
const int LED_BUILTIN2 = 16; //green
const int LED_BUILTIN3 = 17; //blue
const int voltage_read = 35;
const int illum_read = 34;
const int blink_speed = 10000; //read every 10sec

float batteryFraction;
float voltageLevel;
int vbat;
int illum;

// Implement heartbeat led
const int fadeTime = 1000; 
unsigned long lastMillis = 0;
int brightness = 0;
int fadeDirection = 1;  // 1 for fading in, -1 for fading out

//Voltage read
const int MAX_ANALOG_VAL = 4095;
const float MAX_BATTERY_VOLTAGE = 4.2; // Max LiPoly voltage of a 3.7 battery is 4.2

void custom_setup()
{
    // Initialization code here
    last_blink = millis();
    //pinMode(LED_BUILTIN1, OUTPUT);
    //digitalWrite(LED_BUILTIN1, LOW);
    //pinMode(LED_BUILTIN2, OUTPUT);
    //digitalWrite(LED_BUILTIN2, LOW);
    //pinMode(LED_BUILTIN3, OUTPUT);
    //digitalWrite(LED_BUILTIN3, LOW);   
    pinMode(voltage_read, INPUT);
    pinMode(illum_read, INPUT);
    randomSeed(millis());
}

void custom_loop()
{
    // read voltage every 10 seconds
    if(blink_speed && (millis() - last_blink > blink_speed)) {

        vbat = analogRead(voltage_read);
        illum = analogRead(illum_read);
        voltageLevel = (vbat / 4095.0) * 4.2; // calculate voltage level
        batteryFraction = voltageLevel / MAX_BATTERY_VOLTAGE*100;

        if (batteryFraction <= 15) {
            low_bat_alert(LED_BUILTIN1, fadeTime);
        } 
        last_blink = millis();
    }
}

// Function to update the LED brightness for a heartbeat effect
void low_bat_alert(int pin, int duration) {
  unsigned long currentMillis = millis();

  // Check if it's time to update the LED brightness
  if (currentMillis - lastMillis >= duration / 255) {
    lastMillis = currentMillis;

    // Update brightness based on the fade direction
    brightness += fadeDirection;

    // Change direction if reaching maximum or minimum brightness
    if (brightness >= 255 || brightness <= 0) {
      fadeDirection *= -1;
    }

    analogWrite(pin, brightness);
  }
}

void custom_every_second()
{
    Serial.print("#");
}

void custom_every_5seconds()
{
    LOG_VERBOSE(TAG_CUSTOM, "%d seconds have passsed...", 5);


    // Convert the integer to a string
    String vbatFraction = String(batteryFraction);
    String vbatLevel = String(voltageLevel);
    String illumLevel = String(illum);

    // Create the JSON string
    String jsonString = "{\"vbat_Fraction\":" + vbatFraction + "}";
    String jsonString2 = "{\"vbat_Level\":" + vbatLevel + "}";
    String jsonString3 = "{\"illum_Level\":" + illumLevel + "}";

    // Convert the JSON string to a const char* for your function
    const char* jsonChar = jsonString.c_str();
    const char* jsonChar2 = jsonString2.c_str();
    const char* jsonChar3 = jsonString3.c_str();


    // Call your function with the JSON string

    dispatch_state_subtopic("vbat_Fraction", jsonChar);
    dispatch_state_subtopic("vbat_Level", jsonChar2);  
    dispatch_state_subtopic("illum_Level", jsonChar3); 
    
}


bool custom_pin_in_use(uint8_t pin)
{
    
    switch(pin) {
        case illum_read:  // Custom LED pin
        case 6:  // Custom Input pin
            return true;
        default:
            return false;
    }
    
}

void custom_get_sensors(JsonDocument& doc)
{
    JsonObject sensor = doc.createNestedObject(F("Temperature"));  // Add Key
    sensor[F("Temperature")] = random(256);                        // Set Value
}

void custom_topic_payload(const char* topic, const char* payload, uint8_t source){
    // Not used
}

#endif // HASP_USE_CUSTOM


my_custom.h file:

/* MIT License - Copyright (c) 2019-2022 Francis Van Roie
   For full license information read the LICENSE file in the project folder */

// USAGE: - Copy this file and rename it to my_custom.h
//        - uncomment in your user_config_override.h the line containing #define HASP_USE_CUSTOM 1
//

#ifndef HASP_CUSTOM_H
#define HASP_CUSTOM_H

#include "hasplib.h"
#if defined(HASP_USE_CUSTOM)

/* This function is called at boot */
void custom_setup();

/* This function is called every itteration of the main loop */
void custom_loop();

/* This function is called every second */
void custom_every_second();

/* This function is called every 5 seconds */
void custom_every_5seconds();

/* This function is called when battery voltage is below 15% */
void low_bat_alert(int pin, int duration);

/* return true if the pin used by the custom code */
bool custom_pin_in_use(uint8_t pin);

/* Add a key which defines a JsonObject to add to the sensor JSON output */
void custom_get_sensors(JsonDocument& doc);

/* Receive custom topic & payload messages */
void custom_topic_payload(const char* topic, const char* payload, uint8_t source);

#endif // HASP_USE_CUSTOM

#endif // HASP_CUSTOM_H

**Reply by richardsg307** [reply]
This is good stuff but as a complete novice I got lost as soon as I entered the GitPod website.

I have a Guition 4848S040 which makes available GPIOs 1.2 and 40 on a connector so I'm guessing Pins 1 or 2 would do for the Battery divider input. I'd want to display the battery percentage on page 0 and also make it available to Home Assistant through its OpenHasp Integration. I have a 2500mAh battery connected.

Currently I have it loaded with the Guition 4848S040 build from the Nightly Build website.

Using your example, I think I can code the custom file, but any guidance on how to use the GitPod would be welcomed.

Update: I think I've done all the coding, but being unfamiliar with Unix and GitPod, I don't know how to save the template files which I have edited!

**Reply by richardsg307** [reply]
I've managed to compile my code now and flash to my device. But now I'm stuck on how to display it.

My code fragment is:
```
void custom_every_5seconds()
{
    LOG_VERBOSE(TAG_CUSTOM, "5 seconds have passsed...");
    String vbatFraction = String(batteryFraction);
    String vbatLevel = String(voltageLevel);
 
    String jsonStringL = "{\"vbat_Level\":" + vbatLevel + "}";
    String jsonStringF = "{\"vbat_Fraction\":" + vbatFraction + "}";
    const char* jsonCharL = jsonStringL.c_str();
    const char* jsonCharF = jsonStringF.c_str();

    dispatch_state_subtopic("vbat_Fraction", jsonCharF);
    dispatch_state_subtopic("vbat_Level", jsonCharL);
    
}
```

I don't really know what this code does exactly, but I'm guessing it makes eg the battery level appear internally as vbat_Level and is a Float.

I tried to use it as a "val": vbat_Level, and various other syntaxes but nothing seems to work.

**Reply by fvanroie** [reply]
Please see #928
